tests/integration/pilot/mirror_test.go

```diff
@@ -0,0 +1,248 @@
+					vsc := VirtualServiceMirrorConfig{
+						c.name,
+						ns.Name(),
+						c.absent,
+						c.percentage,
+					}
+
+					deployment := tmpl.EvaluateOrFail(t,
+						file.AsStringOrFail(t, "testdata/traffic-mirroring-template.yaml"), vsc)
+					g.ApplyConfigOrFail(t, ns, deployment)
+					defer g.DeleteConfigOrFail(t, ns, deployment)
+
+					workloads, err := instances[0].Workloads()
+					if err != nil {
+						t.Fatalf("Failed to get workloads. Error: %v", err)
+					}
+
+					for _, w := range workloads {
+						if err = w.Sidecar().WaitForConfig(func(cfg *envoyAdmin.ConfigDump) (bool, error) {
+							validator := structpath.ForProto(cfg)
+							if err = checkIfMirrorWasApplied(instances[1], instances[2], c, validator); err != nil {
+								return false, err
+							}
+							return true, nil
+						}); err != nil {
+							t.Fatalf("Failed to apply configuration. Error: %v", err)
+						}
+					}
+
+					testID := util.RandomString(16)
+					sendTrafficMirror(t, instances, testID)
+					verifyTrafficMirror(t, instances, c, testID)
+				})
+			}
+		})
+}
+
+func checkIfMirrorWasApplied(target, mirror echo.Instance, tc testCaseMirror, validator *structpath.Instance) error {
+	for _, port := range target.Config().Ports {
+		vsName := vsName(target, port)
+		instance := validator.Select(
+			"{.configs[*].dynamicRouteConfigs[*].routeConfig.virtualHosts[?(@.name == %q)].routes[*].route}",
+			vsName)
+
+		if tc.percentage > 0 {
+			instance.Exists("{.requestMirrorPolicy}")
+
+			clusterName := clusterName(mirror, port)
+			instance.Equals(clusterName, "{.requestMirrorPolicy.cluster}")
+
+			instance.Equals(tc.percentage, "{.requestMirrorPolicy.runtimeFraction.defaultValue.numerator}")
+		} else {
+			instance.NotExists("{.requestMirrorPolicy}")
+		}
+
+		if err := instance.Check(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func vsName(target echo.Instance, port echo.Port) string {
+	cfg := target.Config()
+	return fmt.Sprintf("%s.%s.svc.%s:%d", cfg.Service, cfg.Namespace.Name(), cfg.Domain, port.ServicePort)
+}
+
+func sendTrafficMirror(t *testing.T, instances [3]echo.Instance, testID string) {
+	const totalThreads = 10
+	errs := make(chan error, totalThreads)
+
+	wg := sync.WaitGroup{}
+	wg.Add(totalThreads)
+
+	for i := 0; i < totalThreads; i++ {
+		go func() {
+			_, err := instances[0].Call(echo.CallOptions{
+				Target:   instances[1],
+				PortName: "http",
+				Path:     "/" + testID,
+				Count:    50,
+			})
+			if err != nil {
+				errs <- err
+			}
+			wg.Done()
+		}()
+	}
+
+	wg.Wait()
+	close(errs)
+
+	var callerrors error
+	for err := range errs {
+		callerrors = multierror.Append(err, callerrors)
+	}
+	if callerrors != nil {
+		t.Fatalf("Error occurred during call: %v", callerrors)
+	}
+}
+
+func verifyTrafficMirror(t *testing.T, instances [3]echo.Instance, tc testCaseMirror, testID string) {
+	_, err := retry.Do(func() (interface{}, bool, error) {
+		countB := logCount(t, instances[1], testID)
```

Comment:
Is this counting the number of logs in the container and relying on access logs as a proxy for number of requests? I guess we cannot do what we do in trafficshifting/locality test of looking at the response `host` since this is mirroring?

Reply:
We make a request from A to B, the response we get comes from B, not from C (which got the mirrored requests). We need to measure the responses from C in order to check if the percentage is correct. AFAIK there's no way of doing that from the response we get from B.